File: C++/deleteDuplicates2.cpp
#include<iostream>

struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
  public:
  ListNode* deleteDuplicates(ListNode* head) {
       ListNode *dummy = new ListNode(-1), *cur = dummy;
        dummy->next = head;
        while (cur->next && cur->next->next) {
            if (cur->next->val == cur->next->next->val) {
                int tmp = cur->next->val;
                while (cur->next && cur->next->val == tmp) {
                    cur->next = cur->next->next;
                }
            }
            else cur = cur->next;
        }
        return dummy->next;
    }
};

int main(){
    ListNode* head=new ListNode(1);
    head->next=new ListNode(2);
    head->next->next=new ListNode(3);
    head->next->next->next=new ListNode(3);
    head->next->next->next->next=new ListNode(4);
    head->next->next->next->next->next=new ListNode(4);
    head->next->next->next->next->next->next=new ListNode(5);
    Solution solution;

    std::cout<<"Original list:";
    for(ListNode* node=head;node !=nullptr;node=node->next){
        std::cout<<node->val<<" ";
    }
    std::cout<<std::endl;

    ListNode* deleteHead=solution.deleteDuplicates(head);
    while(deleteHead){
        std::cout<<deleteHead->val<<" ";
        deleteHead=deleteHead->next;
    }
    std::cout<<std::endl;

    while(head){
        ListNode* next=head->next;
        delete head;
        head=next;
    }
    return 0;
}

File: python/deleteDuplicates2.py
class ListNode:
    def __init__(self, x):
            self.val = x
            self.next = None

class Solution:
    def deleteDuplicates(self , head: ListNode) -> ListNode:
        if head == None: 
            return None
        res = ListNode(0)
        res.next = head 
        cur = res
        while cur.next and cur.next.next: 
            if cur.next.val == cur.next.next.val: 
                temp = cur.next.val
                while cur.next != None and cur.next.val == temp: 
                    cur.next = cur.next.next
            else:
                cur = cur.next
        return res.next 

head=ListNode(1)
head.next=ListNode(2)
head.next.next=ListNode(3)
head.next.next.next=ListNode(3)
head.next.next.next.next=ListNode(4)
head.next.next.next.next.next=ListNode(4)
head.next.next.next.next.next.next=ListNode(5)
solution=Solution()
delete_duphead=solution.deleteDuplicates(head)
while delete_duphead:
    print(delete_duphead.val,end='')
    delete_duphead=delete_duphead.next
print()